sregistry/main/s3/__init__.py

```diff
@@ -79,18 +81,29 @@ def get_bucket(self):
 
         self.bucket = self.s3.Bucket(self.bucket_name)
         # check if the bucket exists and can be accessed
+        # based on https://stackoverflow.com/a/47565719/1253230
         try:
             self.s3.meta.client.head_bucket(Bucket=self.bucket.name)
-        except ClientError:
+        except ClientError as e:
             self.bucket = None
- 
-        # If the bucket doesn't exist, try creating it
-        if self.bucket is None:
-            try:
-                self.bucket = self.s3.create_bucket(Bucket=self.bucket_name)
-                bot.info('Created bucket %s' % self.bucket.name )
-            except ClientError as e:
-                bot.exit('Could not create bucket {}: {}'.format(self.bucket_name, str(e)))
+            error_code = int(e.response['Error']['Code'])
+            if error_code == 403:
+                # print("Private Bucket. Forbidden Access!")
+                if not self._id or not self._key:
+                    bot.exit('The {} bucket is not public and needs AWS credentials to be accessed'.format(self.bucket_name))
+                bot.exit('The {} bucket cannot be accessed with the provided AWS credentials'.format(self.bucket_name))
+            elif error_code == 404:
+                # If the bucket doesn't exist, ask the user if he/she wants to try creating it
+                if confirm_action('Are you sure you want to create the {} bucket?'.format(self.bucket_name)):
+                    if not self._id or not self._key:
+                        bot.exit('Buckets need AWS credentials to be created')
+                    try:
+                        self.bucket = self.s3.create_bucket(Bucket=self.bucket_name)
+                        bot.info('Created bucket %s' % self.bucket.name )
+                    except ClientError as e:
+                        bot.exit('Could not create bucket {}: {}'.format(self.bucket_name, str(e)))
+                else:
+                    bot.exit('Aborting due to not creating bucket')
 
         return self.bucket
```

Comment:
So when we get here - there is absolutely not a chance that we don't have a bucket?

Reply:
Hmmm... I guess I haven't dealt with the case of a different type of exception (not sure if that's even possible, but this would bring down the whole program immediately), as well as the case of the error code not being 404 or 403, in which case the bucket is `None`, which will lead to an unhandled `TypeError` later on.

Reply:
Do you think it would be safe to add an else for the error code (in case not 404/403) with something like "Unrecognized error code X, exiting." We need to make sure that we don't get to the bottom and return something given that there was an issue.

Reply:
Fixed! I also did a bit interactive rebasing to clean up my commits...

Reply:
This looks much better. If you've tested and are happy, I'm good to merge. If it's imperfect or some other issue comes up in the future, the user can report and we will fix.